File: app/src/main/java/org/apache/cordova/Condo.kt
package org.apache.cordova

import ai.doma.core.DI.CoreComponent
import ai.doma.core.DI.CoreModule
import ai.doma.miniappdemo.collectAndTrace
import ai.doma.miniappdemo.data.RetrofitApi
import ai.doma.miniappdemo.domain.MiniappBackStack
import ai.doma.miniappdemo.domain.MiniappBackStackEntry
import ai.doma.miniappdemo.ext.logD
import android.webkit.ValueCallback
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.cancel
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import okhttp3.FormBody
import okhttp3.Interceptor
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.net.URLEncoder
import java.nio.Buffer
import java.util.*
object CondoPluginState {
    var payload: String? = null
}
class Condo : CordovaPlugin() {
    private var authState: String? = null

    var api: RetrofitApi? = null
    init {
        api = CoreComponent.get()?.retrofitApi
        logD { "CondoPlugin init api: $api" }
    }

    val scope = CoroutineScope(Dispatchers.IO)
    protected fun finalize() {
        scope.cancel()
    }

    @Throws(JSONException::class)
    override fun execute(
        action: String,
        args: JSONArray,
        callbackContext: CallbackContext
    ): Boolean {
        if (action == ACTION_REQUEST_AUTH_CODE) {
            scope.launch {
                requestAuthCode(clientID = args.getString(0))
                    .collectAndTrace {
                        callbackContext.success(it)
                    }
            }
        }
        if (action == ACTION_REQUEST_AUTH) {
            scope.launch {
                requestAuthCode(clientID = args.getString(0))
                    .flatMapLatest { code ->
                        requestAuth(
                            code = code,
                            clientID = args.getString(0),
                            clientSecret = args.getString(1)
                        )
                    }
                    .collectAndTrace {
                        callbackContext.success(it)
                    }
            }
        }
        if (action == ACTION_REQUEST_SERVER_AUTH) {
            scope.launch {
                requestFullAuth(
                    client_id = args.getString(0),
                    redirect_uri = args.getString(1),
                )
                    .collectAndTrace { response ->

                        if(response.isSuccessful){
                            val request = response.raw().request
                            val lastRedirect = request.url.toString()
                            callbackContext.success(lastRedirect)
                        } else {
                            callbackContext.sendPluginResult(PluginResult(PluginResult.Status.ERROR))
                        }
                    }
            }
        }

        if (action == ACTION_REQUEST_SERVER_AUTH_BY_URL) {
            scope.launch {
                requestFullAuthByUrl(url = args.getString(0)).collectAndTrace { response ->

                    if(response.isSuccessful){
                        val request = response.raw().request
                        val lastRedirect = request.url.toString()
                        val lastRedirectBody = okio.Buffer().apply{ request.body?.writeTo(this) }.readUtf8()
                        val json = JSONObject()
                        json.put("url", lastRedirect)
                        json.put("body", lastRedirectBody)
                        json.put("status", lastRedirectBody)
                        callbackContext.success(json)
                    } else {
                        callbackContext.sendPluginResult(PluginResult(PluginResult.Status.ERROR))
                    }


                }
            }
        }
        if (action == ACTION_GET_CURRENT_RESIDENT){
            val testJson = "{\"_label_\":\"37c0f190-3daf-4836-b47e-8251d6004274\",\"address\":\"г Владивосток, Океанский пр-кт, д 20\",\"addressKey\":\"d6af941f-a495-4331-9724-2048bf47ff0a\",\"addressMeta\":{\"data\":{\"capital_marker\":\"2\",\"city\":\"Владивосток\",\"city_district\":\"Ленинский\",\"city_district_type\":\"р-н\",\"city_district_type_full\":\"район\",\"city_district_with_type\":\"Ленинский р-н\",\"city_fias_id\":\"7b6de6a5-86d0-4735-b11a-499081111af8\",\"city_kladr_id\":\"2500000100000\",\"city_type\":\"г\",\"city_type_full\":\"город\",\"city_with_type\":\"г Владивосток\",\"country\":\"Россия\",\"country_iso_code\":\"RU\",\"federal_district\":\"Дальневосточный\",\"fias_actuality_state\":\"0\",\"fias_id\":\"21440ab7-280c-4401-aab6-32f6176ece60\",\"fias_level\":\"8\",\"geo_lat\":\"43.119392\",\"geo_lon\":\"131.88748\",\"geoname_id\":\"2013348\",\"house\":\"20\",\"house_fias_id\":\"21440ab7-280c-4401-aab6-32f6176ece60\",\"house_kladr_id\":\"2500000100003660131\",\"house_type\":\"д\",\"house_type_full\":\"дом\",\"kladr_id\":\"2500000100003660131\",\"okato\":\"05401364000\",\"oktmo\":\"05701000001\",\"postal_code\":\"690091\",\"qc_geo\":\"0\",\"region\":\"Приморский\",\"region_fias_id\":\"43909681-d6e1-432d-b61f-ddac393cb5da\",\"region_iso_code\":\"RU-PRI\",\"region_kladr_id\":\"2500000000000\",\"region_type\":\"край\",\"region_type_full\":\"край\",\"region_with_type\":\"Приморский край\",\"street\":\"Океанский\",\"street_fias_id\":\"d759fede-e5dc-4cf7-9d95-53a851778847\",\"street_kladr_id\":\"25000001000036600\",\"street_type\":\"пр-кт\",\"street_type_full\":\"проспект\",\"street_with_type\":\"Океанский пр-кт\",\"tax_office\":\"2536\",\"tax_office_legal\":\"2536\"},\"unrestricted_value\":\"690091, Приморский край, г Владивосток, Ленинский р-н, Океанский пр-кт, д 20\",\"value\":\"г Владивосток, Океанский пр-кт, д 20\"},\"createdAt\":\"2023-06-26T15:16:48.908Z\",\"createdBy\":{\"_label_\":\"Дзампаев Батраз Олегович -- <d6e45786-9c22-4deb-a011-3440c05c0318>\",\"createdAt\":\"2022-02-24T12:54:09.711Z\",\"dv\":1,\"id\":\"d6e45786-9c22-4deb-a011-3440c05c0318\",\"name\":\"Дзампаев Батраз Олегович\",\"type\":\"resident\",\"updatedAt\":\"2022-08-22T09:39:13.837Z\",\"v\":3},\"dv\":1,\"id\":\"37c0f190-3daf-4836-b47e-8251d6004274\",\"organization\":{\"_label_\":\"Владивосток Топ -- <fb0d418c-f104-44c5-815d-15265c9925f8>\",\"country\":\"ru\",\"dv\":1,\"id\":\"fb0d418c-f104-44c5-815d-15265c9925f8\",\"name\":\"Владивосток Топ\",\"tin\":\"7451405877\",\"type\":\"MANAGING_COMPANY\",\"updatedAt\":\"2024-02-15T13:45:18.854Z\",\"v\":3},\"organizationFeatures\":{\"hasBillingData\":true,\"hasMeters\":true},\"paymentCategories\":[{\"acquiringName\":\"Расчетный Банк\",\"billingName\":\"Реестровый обмен\",\"categoryName\":\"Квартплата\",\"id\":\"1\"},{\"acquiringName\":\"default\",\"billingName\":\"default\",\"categoryName\":\"Электричество\",\"id\":\"3\"},{\"acquiringName\":\"default\",\"billingName\":\"default\",\"categoryName\":\"Газ\",\"id\":\"4\"},{\"acquiringName\":\"default\",\"billingName\":\"default\",\"categoryName\":\"Холодная вода\",\"id\":\"5\"},{\"acquiringName\":\"default\",\"billingName\":\"default\",\"categoryName\":\"Отопление и горячая вода\",\"id\":\"6\"},{\"acquiringName\":\"default\",\"billingName\":\"default\",\"categoryName\":\"Мусор\",\"id\":\"7\"},{\"acquiringName\":\"Расчетный Банк\",\"billingName\":\"Реестровый обмен\",\"categoryName\":\"Капремонт\",\"id\":\"9\"}],\"property\":{\"id\":\"032f84ad-a36f-484b-bc69-a1f806d080f1\",\"name\":\"\"},\"unitName\":\"1\",\"unitType\":\"flat\",\"updatedAt\":\"2023-06-26T15:16:48.908Z\",\"updatedBy\":{\"_label_\":\"Дзампаев Батраз Олегович -- <d6e45786-9c22-4deb-a011-3440c05c0318>\",\"createdAt\":\"2022-02-24T12:54:09.711Z\",\"dv\":1,\"id\":\"d6e45786-9c22-4deb-a011-3440c05c0318\",\"name\":\"Дзампаев Батраз Олегович\",\"type\":\"resident\",\"updatedAt\":\"2022-08-22T09:39:13.837Z\",\"v\":3},\"user\":{\"_label_\":\"Дзампаев Батраз Олегович -- <d6e45786-9c22-4deb-a011-3440c05c0318>\",\"createdAt\":\"2022-02-24T12:54:09.711Z\",\"dv\":1,\"id\":\"d6e45786-9c22-4deb-a011-3440c05c0318\",\"name\":\"Дзампаев Батраз Олегович\",\"phone\":\"+79897456577\",\"type\":\"resident\",\"updatedAt\":\"2022-08-22T09:39:13.837Z\",\"v\":3},\"v\":1}"
            callbackContext.success(JSONObject(testJson))
        }
        if (action == ACTION_CLOSE_MINIAPP) {
            this.cordova.onMessage(ACTION_CLOSE_MINIAPP, Unit)
        }

        if (action == ACTION_GET_LAUNCH_CONTEXT) {
            callbackContext.success(CondoPluginState.payload.orEmpty())
        }
        if (action == ACTION_NOTIFY_CALL_ENDED) {
            callbackContext.success()
        }
        if (action == ACTION_HISTORY_PUSH_STATE) {
            MiniappBackStack.push(MiniappBackStackEntry(args.getString(1), args.get(0)))

            sendHistoryStateToJs()
            callbackContext.success()
        }
        if (action == ACTION_HISTORY_BACK) {
            MiniappBackStack.pop()
            sendHistoryStateToJs()
            callbackContext.success()
        }
        if (action == ACTION_HISTORY_GO) {
            MiniappBackStack.pop(args.getLong(0))
            sendHistoryStateToJs()
            callbackContext.success()
        }
        if (action == ACTION_HISTORY_REPLACE_STATE) {
            MiniappBackStack.replace(MiniappBackStackEntry(args.getString(1), args.get(0)))
            sendHistoryStateToJs()
            callbackContext.success()
        }

        logD { "CondoPlugin execute: action: ${action} args: ${args.toString()}" }
        return true
    }


    private fun requestAuthCode(clientID: String) = flow {
        authState = UUID.randomUUID().toString()
        val redirectURI = "https%3A%2F%2Fmobile.doma.ai"

        emit(
            api?.miniappRequestAuthCode(
                url = "https://condo.d.doma.ai/oidc/auth?response_type=code&client_id=${clientID}&redirect_uri=${redirectURI}&scope=openid&state=${authState}",
                cookie = "keystone.sid=${URLEncoder.encode("s:${CoreModule.access_token}", "UTF-8")}"
            ) ?: throw Exception("no retrofit instance found")
        )
    }.map {
        it.raw().request.url.queryParameter("code") ?: ""
    }

    private fun requestAuth(code: String, clientID: String, clientSecret: String) = flow {
        authState = UUID.randomUUID().toString()

        val testAuthHeader = "Basic ${Base64.getEncoder().encodeToString("$clientID:$clientSecret".toByteArray())}"

        val body = FormBody.Builder()
            .addEncoded("grant_type", "authorization_code")
            .addEncoded("code", code)
            .addEncoded("redirect_uri", "https%3A%2F%2Fmobile.doma.ai")
            .build()

        emit(
            api?.miniappRequestAuth(
                url = "https://condo.d.doma.ai/oidc/token",
                authorizationHeader = testAuthHeader,
                body = body
            ) ?: throw Exception("no retrofit instance found")
        )
    }.map {
        it.string()
    }



    private fun requestFullAuth(client_id: String, redirect_uri: String) = flow {
        authState = UUID.randomUUID().toString()
        val url = "https://condo.d.doma.ai/oidc/auth?response_type=code&client_id=${client_id}&redirect_uri=${redirect_uri}&scope=openid&state=${authState})"
        val cookie = "keystone.sid=${URLEncoder.encode("s:${CoreModule.access_token}", "UTF-8")}"
        emit(api?.miniappRequestFullAuth(url, cookie) ?: throw Exception("no retrofit instance found"))
    }

    private fun requestFullAuthByUrl(url: String) = flow {
        val cookie = "keystone.sid=${URLEncoder.encode("s:${CoreModule.access_token}", "UTF-8")}"
        emit(api?.miniappRequestFullAuth(url, cookie) ?: throw Exception("no retrofit instance found"))
    }

    fun sendHistoryStateToJs(){
        val state = MiniappBackStack.backstack.value.lastOrNull()?.state
        val stateStr = if (state is String){
            """"$state""""
        } else state.toString()
        scope.launch(Dispatchers.Main) {

            this@Condo.webView.engine.evaluateJavascript("""window.dispatchEvent(new PopStateEvent('condoPopstate', { 'state': $stateStr}));""", ValueCallback {
                logD { it }
            })
        }
    }


    companion object {
        private const val ACTION_REQUEST_AUTH_CODE = "requestAuthorizationCode"
        private const val ACTION_REQUEST_AUTH = "requestAuthorization"
        private const val ACTION_REQUEST_SERVER_AUTH = "requestServerAuthorization"
        private const val ACTION_REQUEST_SERVER_AUTH_BY_URL = "requestServerAuthorizationByUrl"
        private const val ACTION_GET_CURRENT_RESIDENT = "getCurrentResident"
        private const val ACTION_GET_LAUNCH_CONTEXT = "getLaunchContext"
        private const val ACTION_NOTIFY_CALL_ENDED = "notifyCallEnded"


        private const val ACTION_HISTORY_PUSH_STATE = "historyPushState"
        private const val ACTION_HISTORY_BACK = "historyBack"
        private const val ACTION_HISTORY_REPLACE_STATE = "historyReplaceState"
        private const val ACTION_HISTORY_GO = "historyGo"

        const val ACTION_CLOSE_MINIAPP = "closeApplication"
    }


}

File: MainCordovaApplication/plugins/cordova-plugin-condo/src/ios/Condo.m
/********* Condo.m Cordova Plugin Implementation *******/

#import <Cordova/CDV.h>
#ifdef DOMA
#import "CordovaDemoApp-Swift.h"
#endif

@interface Condo: CDVPlugin 
{
  // Member variables go here.
}
- (void)requestAuthorization:(CDVInvokedUrlCommand *)command;
- (void)closeApplication:(CDVInvokedUrlCommand *)command;
- (void)getCurrentResident:(CDVInvokedUrlCommand *)command;
- (void)requestServerAuthorizationByUrl:(CDVInvokedUrlCommand *)command;
- (void)openURLWithFallback:(CDVInvokedUrlCommand *)command;
#ifdef DOMA
@property (nonatomic, strong) DemoApplicationMainAppAPI *api;
#endif
@end

@implementation Condo

- (void)requestAuthorization:(CDVInvokedUrlCommand *)command
{
    NSString *clientId = [command.arguments objectAtIndex:0];
    NSString *clientSecret = [command.arguments objectAtIndex:1];
    
    if (clientId != nil && [clientId isKindOfClass:[NSString class]] && [clientId length] > 0) 
    {
#ifdef DOMA
        if (self.api == nil)
        {
            self.api = [[DemoApplicationMainAppAPI alloc] init];
        }
        [self.api getCodeWithClientID:clientId complition:^(NSString *code)
        {
            if (code != nil && [code isKindOfClass:[NSString class]] && [code length] > 0)
            {
                if (self.api == nil)
                {
                    self.api = [[DemoApplicationMainAppAPI alloc] init];
                }
                [self.api sendCodeToServerWithCode:code clientId:clientId clientSecret:clientSecret complition:^(NSString *authorization) {
                    if (authorization != nil && [authorization isKindOfClass:[NSString class]] && [authorization length] > 0)
                    {
                        [self.commandDelegate sendPluginResult:[CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:authorization] callbackId:command.callbackId];
                    }
                    else
                    {
                        [self.commandDelegate sendPluginResult:[CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR] callbackId:command.callbackId];
                    }
                }];
            }
            else
            {
                [self.commandDelegate sendPluginResult:[CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR] callbackId:command.callbackId];
            }
        }];
#else
        [self.commandDelegate sendPluginResult:[CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@"test_code_for_auth"] callbackId:command.callbackId];
#endif
    } 
    else 
    {
        [self.commandDelegate sendPluginResult:[CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR] callbackId:command.callbackId];
    }
}

- (BOOL)nonEmptyString:(NSString *)string
{
    return (string != nil && [string isKindOfClass:[NSString class]] && [string length] > 0);
}

- (void)requestServerAuthorizationByUrl:(CDVInvokedUrlCommand *)command
{
    NSString *url = [command.arguments objectAtIndex:0];
//    reserved for future use
//    NSString *custom_params = [command.arguments objectAtIndex:1];
    
    if ([self nonEmptyString:url])
    {
#ifdef DOMA
        if (self.api == nil)
        {
            self.api = [[DemoApplicationMainAppAPI alloc] init];
        }
        [self.api getFullAuthByUrlWithUrl:url complition:^(NSDictionary *result) {
            [self.commandDelegate sendPluginResult:[CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:result] callbackId:command.callbackId];
        }];
#else
        [self.commandDelegate sendPluginResult:[CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@"test_code_for_auth"] callbackId:command.callbackId];
#endif
    }
    else
    {
        [self.commandDelegate sendPluginResult:[CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR] callbackId:command.callbackId];
    }
}

- (void)openURLWithFallback:(CDVInvokedUrlCommand *)command
{
    NSString *url = [command.arguments objectAtIndex:0];
    NSString *fallback_url = [command.arguments objectAtIndex:1];
    if ([self nonEmptyString:url])
    {
        NSURL *firstURL = [NSURL URLWithString:url];
        [[UIApplication sharedApplication] openURL:firstURL options:@{} completionHandler:^(BOOL success) {
            if (!success)
            {
                NSURL *fallbackURL = [NSURL URLWithString:fallback_url];
                [[UIApplication sharedApplication] openURL:fallbackURL options:@{} completionHandler:^(BOOL success) {}];
            }
        }];
    }
    [self.commandDelegate sendPluginResult:[CDVPluginResult resultWithStatus:CDVCommandStatus_OK] callbackId:command.callbackId];
}

- (void)closeApplication:(CDVInvokedUrlCommand *)command
{
#ifdef DOMA
    [[CondoSupport getInstance] closeApp];
#endif
    [self.commandDelegate sendPluginResult:[CDVPluginResult resultWithStatus:CDVCommandStatus_OK] callbackId:command.callbackId];
}

- (void)getCurrentResident:(CDVInvokedUrlCommand *)command
{
    NSDictionary *dic = [NSJSONSerialization JSONObjectWithData:[@"{\"unitName\":\"1\",\"dv\":1,\"property\":{\"id\":\"94e2f48c-24b3-4f0d-bc9d-74ab863805a1\",\"name\":\"Лермонтова, д 7\"},\"organizationName\":\"Тестирую API\",\"organizationFeatures\":{\"hasBillingData\":false,\"hasMeters\":true},\"createdAt\":\"2022-06-29T18:31:56.786+0500\",\"_label_\":\"b7d77f24-9ae6-4f78-9915-6645175e50d1\",\"organization\":{\"sender\":{\"dv\":1},\"_label_\":\"Тестирую API -- <9c2f63c8-9288-4942-b24f-e11da26f6bec>\",\"country\":\"ru\",\"name\":\"Тестирую API\",\"id\":\"9c2f63c8-9288-4942-b24f-e11da26f6bec\",\"v\":1,\"updatedAt\":\"2022-04-18T15:23:45.431+0500\",\"tin\":\"6671095805\",\"__typename\":\"Organization\",\"dv\":1},\"v\":2,\"paymentCategories\":[{\"id\":\"1\",\"acquiringName\":\"default\",\"billingName\":\"default\",\"categoryName\":\"Квартплата\"},{\"id\":\"9\",\"acquiringName\":\"default\",\"billingName\":\"default\",\"categoryName\":\"Капремонт\"}],\"updatedAt\":\"2022-07-27T17:37:52.842+0500\",\"id\":\"b7d77f24-9ae6-4f78-9915-6645175e50d1\",\"createdBy\":{\"sender\":{\"dv\":1},\"id\":\"f919b741-1fd4-40f4-99f0-4f3d9ec031ec\",\"v\":5,\"updatedAt\":\"2022-06-29T16:01:49.279+0500\",\"createdAt\":\"2022-06-20T05:39:00.382+0500\",\"type\":\"resident\",\"name\":\"Kghv\",\"dv\":1},\"propertyId\":\"94e2f48c-24b3-4f0d-bc9d-74ab863805a1\",\"unitType\":\"flat\",\"sender\":{\"dv\":1},\"user\":{\"sender\":{\"dv\":1},\"id\":\"f919b741-1fd4-40f4-99f0-4f3d9ec031ec\",\"v\":5,\"updatedAt\":\"2022-06-29T16:01:49.279+0500\",\"createdAt\":\"2022-06-20T05:39:00.382+0500\",\"type\":\"resident\",\"name\":\"Kghv\",\"dv\":1},\"updatedBy\":{\"sender\":{\"dv\":1},\"id\":\"f919b741-1fd4-40f4-99f0-4f3d9ec031ec\",\"v\":5,\"updatedAt\":\"2022-06-29T16:01:49.279+0500\",\"createdAt\":\"2022-06-20T05:39:00.382+0500\",\"type\":\"resident\",\"name\":\"Kghv\",\"dv\":1},\"organizationId\":\"9c2f63c8-9288-4942-b24f-e11da26f6bec\",\"propertyName\":\"Лермонтова, д 7\",\"addressMeta\":{\"value\":\"г Новосибирск, ул Лермонтова, д 7\",\"data\":{\"cityTypeFull\":\"город\",\"region\":\"Новосибирская\",\"taxOfficeLegal\":\"5406\",\"regionIsoCode\":\"RU-NVS\",\"street\":\"Лермонтова\",\"country\":\"Россия\",\"taxOffice\":\"5406\",\"cityWithType\":\"г Новосибирск\",\"regionWithType\":\"Новосибирская обл\",\"oktmo\":\"50701000001\",\"capitalMarker\":\"2\",\"streetFiasId\":\"25327053-e12e-4e6a-ba9e-a25ee594c281\",\"qcGeo\":\"2\",\"regionTypeFull\":\"область\",\"geoLat\":\"55.044805\",\"countryIsoCode\":\"RU\",\"streetType\":\"ул\",\"postalCode\":\"630091\",\"cityFiasId\":\"8dea00e3-9aab-4d8e-887c-ef2aaa546456\",\"cityKladrId\":\"5400000100000\",\"geonameId\":\"1496747\",\"streetWithType\":\"ул Лермонтова\",\"cityType\":\"г\",\"okato\":\"50401386000\",\"federalDistrict\":\"Сибирский\",\"city\":\"Новосибирск\",\"houseType\":\"д\",\"streetTypeFull\":\"улица\",\"houseTypeFull\":\"дом\",\"fiasId\":\"2cac22cd-721c-4f5a-b25b-97a771da7ef4\",\"regionFiasId\":\"1ac46b49-3209-4814-b7bf-a509ea1aecd9\",\"fiasLevel\":\"8\",\"regionKladrId\":\"5400000000000\",\"house\":\"7\",\"regionType\":\"обл\",\"fiasActualityState\":\"0\",\"geoLon\":\"82.926862\",\"kladrId\":\"5400000100007340013\",\"houseFiasId\":\"2cac22cd-721c-4f5a-b25b-97a771da7ef4\",\"streetKladrId\":\"54000001000073400\",\"houseKladrId\":\"5400000100007340013\"},\"unrestrictedValue\":\"630091, Новосибирская обл, г Новосибирск, Центральный р-н, ул Лермонтова, д 7\"},\"address\":\"г Новосибирск, ул Лермонтова, д 7\"}" dataUsingEncoding:NSUTF8StringEncoding] options:0 error:nil];
    [self.commandDelegate sendPluginResult:[CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:dic] callbackId:command.callbackId];
}

@end
